cxml/xml.go

```diff
@@ -0,0 +1,222 @@
+		}
+		switch e := token.(type) {
+		case xml.StartElement:
+			parentNodes.Push(e)
+		case xml.EndElement:
+			parentNodes.Pop()
+		}
+	}
+	encoder.Flush()
+	return nil
+}
+
+// SetAttribute is XML helper that allows to set attribute on a node
+func SetAttribute(e xml.StartElement, name, value string) xml.StartElement {
+	if len(e.Attr) != 0 {
+		for i := range e.Attr {
+			if e.Attr[i].Name.Local == name {
+				e.Attr[i].Value = value
+				return e
+			}
+		}
+	} else {
+		e.Attr = append(e.Attr, xml.Attr{Name: xml.Name{Local: name}, Value: value})
+	}
+	return e
+}
+
+// ReplaceCDATAIf replaces CDATA value of the matched node
+// if the parent node name matches the name
+func ReplaceCDATAIf(val []byte, cond ConditionFunc) TransformFunc {
+	return func(parents *NodeList, in xml.Token) []xml.Token {
+		switch in.(type) {
+		case xml.CharData:
+			if cond(parents, in) {
+				data := make([]byte, len(val))
+				copy(data, val)
+				return []xml.Token{xml.CharData(data)}
+			}
+		}
+		return []xml.Token{in}
+	}
+}
+
+// InjectNodes injects nodes at the end of the tag that matches name
+func InjectNodesIf(nodes []xml.Token, cond ConditionFunc) TransformFunc {
+	return func(parents *NodeList, in xml.Token) []xml.Token {
+		if cond(parents, in) {
+			return append(nodes, in)
+		}
+		return []xml.Token{in}
+	}
+}
+
+// ReplaceAttribute replaces the attribute of the first node that matches the name
+func ReplaceAttributeIf(attrName, attrValue string, cond ConditionFunc) TransformFunc {
+	return func(parents *NodeList, in xml.Token) []xml.Token {
+		switch t := in.(type) {
+		case xml.StartElement:
+			if cond(parents, in) {
+				e := xml.StartElement(t)
+				return []xml.Token{SetAttribute(e, attrName, attrValue)}
+			}
+		}
+		return []xml.Token{in}
+	}
+}
+
+// TrimSpace is a transformer function that replaces CDATA with blank
+// characters with empty strings
+func TrimSpace(parents *NodeList, in xml.Token) []xml.Token {
+	switch t := in.(type) {
+	case xml.CharData:
+		if strings.TrimSpace(string(t)) == "" {
+			return []xml.Token{xml.CharData("")}
+		}
+	}
+	return []xml.Token{in}
+}
+
+// Combine takes a list of TransformFuncs and converts them
+// into a single transform function applying all functions one by one
+func Combine(funcs ...TransformFunc) TransformFunc {
+	return func(parents *NodeList, in xml.Token) []xml.Token {
+		out := []xml.Token{in}
+		for _, f := range funcs {
+			new := []xml.Token{}
+			for _, t := range out {
+				new = append(new, f(parents, t)...)
+			}
+			out = new
+		}
+		return out
+	}
+}
+
+// ParentIs is a functon that returns ConditionFunc checking
+// if immediate parent's name matches name
+func ParentIs(name xml.Name) ConditionFunc {
+	return func(parents *NodeList, el xml.Token) bool {
+		return parents.ParentIs(name)
+	}
+}
+
+// GetAttribute returns attribute value if it's present, otherwise
+// it returns an empty string
+func GetAttribute(e xml.StartElement, name xml.Name) string {
+	if len(e.Attr) == 0 {
```

Comment:
you know what? for some reason I had misconception that range would work on empty list, but will panic on nil list, however that's actually not true:

http://play.golang.org/p/veeLtBdKTO

That's a good catch!